### ISSUE TITLE
SetText hangs indefinitely on Linux

### ISSUE THREAD

**Issue by principis** [CopyText/TextCopy#330]
_I'm not sure how no-one has noticed this yet, but here it is..._

Since #137 was merged, the bashrunner xclip process (when calling `SetText`) never returns until copying something to the clipboard (e.g. pressing `CTRL+C`).

This is expected behavior since the xclip `-loops` option is set to `0` by default and xclip does not close stdout. 
From the man page:
```
-l, -loops
        number of X selection requests (pastes into X applications) to wait for before exiting, with a value of 0 (default)
        causing xclip to wait for an unlimited number of requests until another application (possibly another invocation of
        xclip) takes ownership of the selection
```

The solution seems to be adding `-loops 1` to the command, but I'm not sure if this creates other problems...
A different solutions is using `xsel` since it immediately closes stdout after forking.

**Comment by SimonCropp** [comment]
happy to accept that as a change. but first can u share a failing unit test that illustrates this issue

**Comment by principis** [comment]
> happy to accept that as a change. but first can u share a failing unit test that illustrates this issue

The Snippets tests are failing (because SetTest hangs until I copy something). I'm not sure how I would create unit test specifically for this issue.
I'll create a test application and try it out on different Linux distro's. I'm currently on Fedora 33.

**Comment by principis** [comment]
@SimonCropp 

I've created this application and had some people test it.

It was tested on the following distros:
- Fedora
- Arch

Sadly couldn't test it on debian or ubuntu, but I'll try to update when I find someone.

It hangs (as expected) on the first call, until something is selected or copied (depends on the desktop environment I think)
The other 2 work fine.


```csharp
using System;
using System.Diagnostics;
using TextCopy;

namespace TextCopyTest
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("TextCopy test");
            Console.WriteLine("If the process hangs, copy something to exit.");

            var sw = new Stopwatch();
            sw.Start();
            ClipboardService.SetText("This is a test");
            Console.WriteLine("Copied text: " + ClipboardService.GetText().Trim());
            Console.WriteLine(sw.Elapsed.Seconds + " seconds" + Environment.NewLine);


            Console.WriteLine("xclip with -l 1 test");

            var arguments = "-c \"echo 'This is a test' | xclip -i -loops 1 -selection clipboard\"";
            using var process = new Process {
                StartInfo = new ProcessStartInfo {
                    FileName = "bash",
                    Arguments = arguments,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = false,
                }
            };

            sw.Restart();
            process.Start();

            process.WaitForExit();
            Console.WriteLine("Copied text: " + ClipboardService.GetText().Trim());
            Console.WriteLine(sw.Elapsed.Seconds + " seconds" + Environment.NewLine);

            Console.WriteLine("xsel test");
            arguments = "-c \"echo 'This is a test' | xsel -i --clipboard\"";
            using var process2 = new Process {
                StartInfo = new ProcessStartInfo {
                    FileName = "bash",
                    Arguments = arguments,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = false,
                }
            };

            sw.Restart();
            process2.Start();

            process.WaitForExit();
            Console.WriteLine("Copied text: " + ClipboardService.GetText().Trim());
            Console.WriteLine(sw.Elapsed.Seconds + " seconds" + Environment.NewLine);

            Console.WriteLine("Test finished");
        }
    }
}
```

**Comment by principis** [comment]
@SimonCropp I think it should be fine to change it. Do you want me to make a PR, or should I let more people test it?

**Comment by SimonCropp** [comment]
@principis happy to accept a PR. can u also include a test to verify the change?

**Comment by mcon** [comment]
Whatever you set `-loops` to it's not going to fix things. 

`xclip` works by spawning a long-lived child process to communicate the clipboard contents (according to a quick strace), it looks like this is what ultimately causes the stdout and stderr file handles to not be closed, and as a result this is what `WaitForExit()` is waiting for.

If you set `-loops 1`, then `xclip` will exit pretty swiftly, but then you don't have the background process running that contains your clipboard contents, which defeats the purpose of the call in the first place...

Looking at the initial PR, it's clear whether this change was tested or not - I'm not sure how it would have ever worked (https://github.com/CopyText/TextCopy/pull/137).

Could I suggest that a reasonable compromise might be to not attempt to print stdout and stderr events when using `xclip`?

**Comment by principis** [comment]
> Looking at the initial PR, it's clear whether this change was tested or not - I'm not sure how it would have ever worked (#137).

@mcon I think your DE (KDE, Gnome, ...) clipboard manager probably handles it.

Your solution seems to work fine, but I'm not sure if that is what we want? 
A different solution would be using `xsel` which properly forks and disassociates itself, so we can exit the process without waiting (the xsel process keeps running until a different program takes over the selection). The downside is that it would break backwards compatibility.

**Comment by mcon** [comment]
Just checking it out now I agree, xsel is better behaved, have verified locally that it fixes the tests for Linux and updated the PR. I agree keeping stdout and stderr is preferable :+1:.

Worth noting that e.g. Ubuntu 18.04 doesn't have `xsel` installed by default, but it's in the Ubuntu PPAs, and the error message propagates up through TextCopy, so don't think this is a huge usability blow (certainly it's better than being broken!), but it's certainly backwards incompatible. 

Not sure Gnome vs KDE should have anything to do with this, as `xsel`/`xclip` both use the standard X "selections" under the hood.

**Comment by principis** [comment]
> Not sure Gnome vs KDE should have anything to do with this, as xsel/xclip both use the standard X "selections" under the hood.

Most desktop environments(?) have their own clipboard manager to cache X selections. That's why it worked fine. You can check it with xclip -v, it immediately gets copied when xclip is executed.

**Comment by principis** [comment]
This is fixed with #345.

Thank you @mcon
